Update to 1.14.x

mdcfe opened the pull request:
# EssentialsX for 1.14 progress

## Reporting issues
**Please report any bugs you encounter by [opening a separate issue](https://github.com/EssentialsX/Essentials/issues/new/choose), *not by responding here*.** If you respond here, your bug report will be ignored. Use this PR for discussing the changes included in the PR.

## Download
You can find experimental builds for 1.14 on [Jenkins](https://ci.ender.zone/job/EssentialsX-Experimental/). If reporting a bug on 1.14, please ensure you're using an experimental `-mc114-bXX` build before reporting the bug!

## Progress
- [x] Compile against latest API
  - [x] 1.14-pre5
  - [x] 1.14
  - [x] 1.14.1
  - [x] 1.14.2
  - *1.14.3 (pending release)*
- [x] Update bStats Metrics class to use Gson
  - [x] Ensure bStats reporting works correctly
- [x] Update sign code to support 1.14 signs
  - [x] Ensure 1.14 signs work as expected
- [x] Add support for 1.14 mobs
  - [x] Update `/kittycannon` to pick between legacy Ocelots and Cats
- [x] Add 1.14 enchantments
- [x] Add backwards compatibility for 1.12 and 1.13 in `MobData`
  - [x] Support `Cat.Type` with `Ocelot.Type` fallbacks
  - [x] Support `Villager.Profession` and `Villager.Career` with fallbacks (fixes #7)
  - [x] Move remaining 1.13/1.14 mob data to MobCompat
- [x] Ensure events are marked async as required
- [x] Update `items.json` to support 1.14 materials
  - [x] Update `items.json` to include fallback materials for older versions of the game
  - [x] Update FlatItemDb to support fallback materials

### Post-merge
- Write more futureproof `items.json` generator

mdcfe committed:
Add prerelease support to VersionUtil

mdcfe committed:
Update bStats Metrics to use Gson

mdcfe committed:
Use MaterialUtil to check signs

mdcfe committed:
Compile against 1.14-pre5

mdcfe committed:
Bump to API 1.14

mdcfe committed:
Implement missing FakeWorld and OfflinePlayer methods

mdcfe committed:
Initial cross-version mob compatibility

Incorporates part of https://github.com/EssentialsX/Essentials/pull/2523 - thanks @JRoy.

mdcfe committed:
Update FakeWorld

mdcfe committed:
Remove redundant BABY_FOX

mdcfe committed:
Add Ocelot.Type fallbacks for all MobCompat.CatTypes

mdcfe committed:
Fix plural suffixes for fish and foxes

mdcfe committed:
Update FakeServer

mdcfe committed:
Improve VersionUtil prerelease handling

Actually includes updated regex this time! (I have no idea how I forgot to update it in cfba203981adb307a45093d167050eebb2fff3c9 - I had already written out the regex)

Incorporates part of https://github.com/EssentialsX/Essentials/pull/2523 - thanks @JRoy.

mdcfe committed:
Fix /kittycannon

The most important command of all!

(This hasn't been tested with 1.13 or below yet.)

JRoy committed:
Add 1.14 Enchantments

cyn8 commented:
Compiled and tested, 1.14 signs work exactly as expected.

`Server version: 1.14-R0.1-SNAPSHOT git-Spigot-4d24e36-ba2e5e7 (MC: 1.14)`
`EssentialsX version: 2.16.1.167`

mdcfe committed:
Add Villager variants, pt 1

mdcfe committed:
Add Villager variants, pt 2

mdcfe committed:
Update bStats Metrics to latest version

Now featuring server brand support!

mdcfe committed:
Fix Metrics package name

mdcfe committed:
Fix Metrics plugin name reporting

JRoy committed:
Update Bukkit to 1.14.1 (#2544)

* Update v1_14_1_R01 as latest supported 1.14 build

* Bump to API 1.14.1

mdcfe committed:
Add llama colour support; improve support for old game versions

Fixes issues with /spawnmob on older versions of the game.

(Also add llama colours, which were missing from the 1.11 update.)

mdcfe committed:
Make AfkStatusChangeEvent async

mdcfe committed:
Use MobCompat.CAT throughout MobData

mdcfe committed:
Use MobCompat entries for other 1.11+ mobs

mdcfe committed:
Only make AfkStatusChangeEvent async when off primary thread

drtshock committed:
Allow AFK status test to pass.

mdcfe committed:
Make all events async when not on primary thread

Fixes #2566, fixes #2556, fixes #2545.

mdcfe committed:
Fix checking for attached wall signs

mdcfe commented on Essentials/src/com/earth2me/essentials/utils/MaterialUtil.java:
```diff
@@ -3,6 +3,7 @@
 import org.bukkit.Bukkit;
 import org.bukkit.DyeColor;
 import org.bukkit.Material;
+import org.bukkit.Registry;
```
Unused import

mdcfe commented on Essentials/src/net/ess3/api/events/AfkStatusChangeEvent.java:
```diff
@@ -1,6 +1,7 @@
 package net.ess3.api.events;
 
 import net.ess3.api.IUser;
+import org.bukkit.Bukkit;
```
Unused import

JRoy committed:
Bump Bukkit to 1.14.2 (#2569)

* Bump to API 1.14.2

* Update v1_14_2_R01 as latest supported 1.14 build

mdcfe committed:
Add fallbacks property to ItemData

Allows items.json to specify fallback names for materials on older versions.

Additionally make non-API methods inaccessible, as they shouldn't be needed by other plugins.

mdcfe committed:
Remove unused import in AFK event

mdcfe committed:
Remove unused import in MaterialUtil

mdcfe committed:
Merge branch 'mc/1.14' of https://github.com/EssentialsX/Essentials into mc/1.14

mdcfe committed:
Implement missing OfflinePlayer methods

mdcfe committed:
Merge 1.13 and 1.14 items.json together

In 2.17.0, [essx-scripts genItemsFlat](https://github.com/md678685/essx-scripts/blob/master/src/genItemsFlat.js) is being replaced with [ItemDbGenerator](https://github.com/md678685/ItemDbGenerator). To ease migration, this is a file generated by merging the 1.14 entries from the new generator into the existing 1.13 file. This will be replaced by an items.json generated entirely by ItemDbGenerator eventually.

mdcfe committed:
Re-add header to items.json

mdcfe committed:
Fix items.json merging
